### ISSUE TITLE
Unusual characters in links can lead to "missing character", even if the character has a glyph in the main font

### ISSUE THREAD

**Issue by munzirtaha** [jgm/pandoc#8802]
```
echo "[[ı]]" | pandoc -f mediawiki --pdf-engine=xelatex -o bug.pdf
[WARNING] Missing character: There is no 𝚤 (U+1D6A4) (U+1D6A4) in font [lmroman10-regular]:mapping=
```
The double square brackets which is used for internal links in mediawiki format caused this. This is not U+1D6A5 character. It is **LATIN SMALL LETTER DOTLESS I (U+0131)**. If I used a single square bracket, there is no problem but then it's the wrong syntax.

**Pandoc version**
```
pandoc --version                                                                                                                          
pandoc 3.0.1
Features: +server +lua
Scripting engine: Lua 5.4
```

**Comment by jgm** [comment]
This is an error from LaTeX.  It says you don't have this character in the default font.
To solve it, use `--metadata mainfont="Your Font"`, where "Your Font" names a system font with this glyph.

**Comment by munzirtaha** [comment]
Yes, I know if I choose a font with character U+1D6A4, it works. However, I don't want to use that math character. I want to use the dotless ı character (U+0131) which is available in the default font.

To verify the code point of the char "ı"
```
hb-shape --verbose lmroman10-regular.otf ı
1: (ı)
1: <U+0131=0>
1: [dotlessi=0+278]
```

To verify the default font has the character
```
ttfdump -t cmap lmroman10-regular.otf | grep 0x0131
                Char 0x0131 -> Index 248
```
Though I don't understand why but using **context** instead of **xelatex** would generate the correct character
`echo "[[ı]]" | pandoc -f mediawiki --pdf-engine=xelatex -o bug.pdf`

I am not sure whether this is a latex or pandoc bug but I guess this information is enough for you to decide whether pandoc needs to change the generated .tex for this to work properly.

**Comment by jgm** [comment]
Ah, sorry, I misunderstood the issue.  Well, that is odd. Pandoc is parsing it properly:
```
% echo "[[ı]]" | pandoc -f mediawiki -t native
[ Para
    [ Link
        ( "" , [] , [] ) [ Str "\305" ] ( "\305" , "wikilink" )
    ]
]
```
The LaTeX writer is then producing
```
\url{ı}
```
where again we have code point 305 (0x131).
So I'm baffled as to why LaTeX is producing that message.  There is no U+1D6A4 character in the source we pass to xelatex...

**Comment by jgm** [comment]
Without the `\url{..}` it works fine -- so I suspect this is something to do with the `\url` command. You'll probably have to ask about this on a TeX forum -- which you can do because it can be reproduced with a simple pure tex file.

**Comment by jgm** [comment]
It could be that `\url` expects a percent-encoded URL.

**Comment by munzirtaha** [comment]
I played a little with the default latex template looking for anything related to math, and luckily when I commented out
`\usepackage{unicode-math}`, xelatex doesn't complain about the missing character. So, why does pandoc use unicode-math by default if it can cause such issues?

**Comment by jgm** [comment]
Good find.

This is the first complaint I've heard about `unicode-math`.  We've been using it for at least five years.  Partly because it allows selection of alternative math fonts, and partly because `mathspec` doesn't work with LuaLaTeX.

My understanding is that `unicode-math` should only be affecting things in math mode, not `\url`.  So, maybe it's a bug, or maybe my understanding is wrong.

**Comment by munzirtaha** [comment]
It seems \url switches to math mode and remap some characters as I see in url.sty. I cannot find an answer to why this is needed!

**Comment by munzirtaha** [comment]
According to https://ftp.cc.uoc.gr/mirrors/CTAN/macros/latex/contrib/url/url.pdf

`Before describing how to customize the printing style, it is best to mention something about the unusual implementation of \url. Although the material is textual in nature, and the font specification required is a text-font command, the text is actually typeset in math mode. This allows the context-sensitive linebreaking, but also accounts for the default behavior of ignoring spaces. (Maybe that underlying design will eventually change.)`

**Comment by jgm** [comment]
Seems like a strange design decision! Anyway, I don't see that there's much we can do about this on our end.
You can always use a percent-encoded URL instead of including the dotless i character verbatim.

**Comment by jgm** [comment]
Or, more controversially, we could percent-encode non-ascii characters before passing them to `\url`.
This would prevent the error in your case, but at the cost of less readable URLs in many other cases.  Hard call.

**Comment by munzirtaha** [comment]
> we could percent-encode non-ascii characters

I agree and don't think there is a better solution.

> before passing them to `\url`.

I tested this with `\url{\%C4\%B1}` but instead of showing `ı`, it printed `%C4%B1` as it is! Then, I tried `\href` and it worked! So, why do you chose `\url` instead of `\href`?


> This would prevent the error in your case, but at the cost of less readable URLs in many other cases. Hard call.

Can't you only encode non-ascii characters? e.g. `https://github.com/AquilaIrreale/urlencode `would not encode ascii characters unless you pass  the `-a` option.

**Comment by jgm** [comment]
> Can't you only encode non-ascii characters? 

Yes, my proposal was precisely: "percent-encode non-ascii characters."  But that still leads to less readable URLs for people whose URLs contain non-ascii characters.

> why do you chose \url instead of \href?

It's been like this a long time, so I honestly don't remember. But, off hand: with `\href` you need to duplicate the URL as the link description.  One copy is better than two.  There may also be other reasons, perhaps connected with line breaking of long URLs.  I'd be hesitant to change without being sure that there wouldn't be bad consequences.

**Comment by jgm** [comment]
>  it printed %C4%B1 as it is!

Well, it's a perfectly valid URL this way.

**Comment by munzirtaha** [comment]
> Well, it's a perfectly valid URL this way.

Sure, it's a valid URL but not a valid description of a URL. `\url{%C4%B1} `is expected to print `ı` in the pdf but point to `%C4%B1`. Currently, it's printing `%C4%B1`! URLs has a limitation that it's acceptable to be encoded but the "text of the link" shouldn't be encoded.

**Comment by jgm** [comment]
`\url` is used for rendering raw URLs as links, so I would think that with `\url` the printed form ("text of the link") should be exactly the same as the link.  Thus `\url` is functioning as intended, at least if you stick to ascii encoded URLs.

The only real question for us is when we should use `\url` and when we should just use `\href` (with possibly duplicated text).

The current logic in the LaTeX writer is that if the link description is `Str x` and the result of unescaping the (perhaps percent-encoded) `x` is the same as the result of unescaping the (perhaps percent-encoded) link destination, then we use `\url`.  Otherwise, `\href`.

Given this situation, it might be prudent to put an additional condition on the use of `\url` -- at minimum, no dotless i, but probably to be safe nothing outside of ASCII.

**Comment by munzirtaha** [comment]
> Yes, my proposal was precisely: "percent-encode non-ascii characters." But that still leads to less readable URLs for people whose URLs contain non-ascii characters.

What about percent-encoding only the problematic characters? I only encountred `ı`, what else?

**Comment by tarleb** [comment]
> What about percent-encoding only the problematic characters? I only encountered ı, what else?

I guess there's no easy way to do this, because whether a character is "problematic" very much depends on the language and fonts used.

It's worth adding that *Noto Sans Math* contains the missing character, using it as main font is a possible work-around.

**Comment by tarleb** [comment]
It seems that the LuaTeX font fallback mechanism doesn't work for math characters, so that's not an option.

**Comment by jgm** [comment]
Changed the writer so that `\url` is only used when the URL is all ASCII.
Otherwise we use `\href`.
